### ISSUE TITLE
Adaptive Quality changes flicker in Unity 5.5.0b6

### ISSUE THREAD

**Issue by automan242** [ValveSoftware/the_lab_renderer#13]
There seems to be an issue in Unity 5.5.0b6  where whenever Adaptive Quality changes one step, the whole screen is rendered with distorted projection for one frame.

Btw, fantastic job on The Lab Renderer! I see amazing framerate and also happy to note that image effects work in the new Unity beta.

**Comment by smygarn** [comment]
Well, I got massive flickering in 5.4.1f1 - More advanced scene = more flicker. It would be better to far better to drop the frame.

**Comment by KevinComerford** [comment]
There seems to be a conflict between the renderer and unity versions after 5.4.1f3. I'm trying to figure out what's causing it as well.

**Comment by SmartCarrion** [comment]
Is this the same as the single pass rendering incompatibility issue?

**Comment by KevinComerford** [comment]
No, that's separate. This has to do with the adaptive quality changing scale. For some reason, when the quality shifts to another level the rendered image is incorrectly scaled for one frame and then is normal again.

The single pass stereo doesn't seem to take into account the adaptive quality at all.

**Comment by djsegler** [comment]
Just switched over to Lab Renderer and I notice the flicker issue with 5.4.1f1. Reducing quality settings helps, but isn't consistent (ie what works yesterday doesn't work as well today...).

Should I drop back to a certain build of 5.4 until this has been resolved? If so, which build?

**Comment by KevinComerford** [comment]
any unity build after 5.4.0f3 causes this flickering issue. its caused by the adpative rendering being at the wrong scale for one frame.

**Comment by djsegler** [comment]
Hmmm - that's pretty far back. 

Any expectations on when a potential solution might be identified? Is it a Unity thing or a Lab Renderer thing?

**Comment by eric-napofearth** [comment]
Ugh. I just updated for a particle fix and ran into this bug. The renderer increasingly loses value if Valve isn't at least keeping it compatible with the latest version of Unity.

**Comment by Selkcip** [comment]
So I found a more specific cause. ValveCamera calls [UpdateAdaptiveQuality ](https://github.com/ValveSoftware/the_lab_renderer/blob/master/Assets/TheLabRenderer/Scripts/ValveCamera.cs#L631)from [OnPreCull](https://github.com/ValveSoftware/the_lab_renderer/blob/master/Assets/TheLabRenderer/Scripts/ValveCamera.cs#L523). I haven't looked at the OpenVR/Unity VR code, but it looks like when you adjust renderScale and renderViewportScale one of the VR APIs adjusts the viewport matrix.

For whatever reason, there's no longer enough time between OnPreCull and actually rendering the frame to make this adjustment so we get a stretched fame.

I moved UpdateAdaptiveQuality to the top of [LateUpdate](https://github.com/ValveSoftware/the_lab_renderer/blob/master/Assets/TheLabRenderer/Scripts/ValveCamera.cs#L444) which, though maybe not ideal, fixed the problem. I think the only adverse effect of this is that the quality will change one frame later than if it was in OnPreCull.

tl;dr Move UpdateAdaptiveQuality() from OnPreCull to LateUpdate

**Comment by bddckr** [comment]
@Selkcip Correct fix I guess. I'm using that same fix in [`VRTK_AdaptiveQuality`](https://github.com/thestonefox/VRTK/blob/7a5e58a64d7869af7045b1cca127bb35f11ae40e/Assets/VRTK/Scripts/VRTK_AdaptiveQuality.cs#L322-L335).
If you want to have a look at that, we have a [PR up to fix it for Unity 5.5.0](https://github.com/thestonefox/VRTK/pull/605) (only tested on Unity 5.5.0beta5, I got reports that 5.5.0beta5 and beta 6 have problems with AQ). There's also a ton of code clean ups and a small performance improvement in [this PR](https://github.com/thestonefox/VRTK/pull/589).

**Comment by arthur5005** [comment]
After a long frustrating evening, I ended up here, thanks for the fix @Selkcip. Moral of the story: Always check Github issues.

**Comment by SmartCarrion** [comment]
Sure would be nice if we could get these fixes into the codebase!

**Comment by Selkcip** [comment]
I think the only way that'll happen is if we fork it and do it ourselves. "The CoLab Renderer"

**Comment by xEvoGx** [comment]
It may or may not be this exact issue, but many times AQ doesn't work at all, scales incorrectly (so it 'appears' I'm looking at the lower corner of the camera-view, including double vision) and swims back and forth depending on frame rate. (nauseating)

Definitely in Forward rendering, so the only thing I can think of are post-effects. Thing is I know I've tried scenes without them and it still did this. 

Equally it HAS worked a times but I can't seem to nail down when it will nor when it won't...

Is there any further steps or things to look for when trying to implement AQ, or is this merely a bug?

**Comment by solidearthlabs** [comment]
Setting the unity Game view "Maximize on Play" has a huge impact on the framerate as well.

See my write-up where I compared the frame timing w/ & w/o the game view maximized:
[http://www.solidearthvr.com/tips/vrgraphicsquality/](http://www.solidearthvr.com/tips/vrgraphicsquality/)

**Comment by BHSPitMonkey** [comment]
News from [Unity's patch notes for Unity 5.5.0p4](https://unity3d.com/unity/qa/patch-releases/5.5.0p4):

> VR: Fixed the usage of VRSettings.renderViewportScale in Camera's OnPreCull so that it was not a frame latent on all supporting SDKs. Fixed issues with Valve's Renderer adaptive quality feature.

I skipped this patch and jumped straight to 5.5.1p1, but I can confirm that the bug is no longer present there!

@automan242 - If this fixes the problem for you as well, consider marking this issue as closed. Thanks for reporting it first and helping me find the helpful workarounds here.

**Comment by automan242** [comment]
Confirmed! Thanks for the headsup.
